### ISSUE TITLE
keepX and keepY are inverted?

### ISSUE THREAD

**Issue by lorenzoFabbri** [mixOmicsTeam/mixOmics#140]
------------
🐞   **Describe the bug:**
The plot showing the possible combinations of the parameters `keepX` and `keepY` seems to be inverted. That is, the vector I assign to `keepX` is used for `keepY` in this plot.

------------
🔍 **reprex results from [reproducible example](https://mixomics-users.discourse.group/t/reproducible-example-to-clarify-issues/470) including sessioninfo():**
```
R version 4.1.0 (2021-05-18)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] cowplot_1.1.1    knitr_1.33       tidytext_0.3.1   scales_1.1.1     ggh4x_0.1.2.1    magrittr_2.0.1   tidyr_1.1.3     
 [8] labelled_2.8.0   stringr_1.4.0    tibble_3.1.3     ggpubr_0.4.0.999 dplyr_1.0.7      mixOmics_6.17.15 ggplot2_3.3.5   
[15] lattice_0.20-44  MASS_7.3-54     

loaded via a namespace (and not attached):
 [1] carData_3.0-4        ellipse_0.4.2        assertthat_0.2.1     cellranger_1.1.0     yaml_2.2.1           ggrepel_0.9.1       
 [7] factoextra_1.0.7     pillar_1.6.1         backports_1.2.1      glue_1.4.2           digest_0.6.27        RColorBrewer_1.1-2  
[13] ggsignif_0.6.2       colorspace_2.0-2     htmltools_0.5.1.9006 Matrix_1.3-4         plyr_1.8.6           pkgconfig_2.0.3     
[19] broom_0.7.8          haven_2.4.1          purrr_0.3.4          corpcor_1.6.9        RSpectra_0.16-0      openxlsx_4.2.4      
[25] rio_0.5.27           BiocParallel_1.26.1  farver_2.1.0         generics_0.1.0       car_3.0-11           ellipsis_0.3.2      
[31] withr_2.4.2          cli_3.0.1            crayon_1.4.1         readxl_1.3.1         evaluate_0.14        tokenizers_0.2.1    
[37] janeaustenr_0.1.5    fansi_0.5.0          SnowballC_0.7.0      rstatix_0.7.0        forcats_0.5.1        foreign_0.8-81      
[43] tools_4.1.0          data.table_1.14.0    hms_1.1.0            lifecycle_1.0.0      matrixStats_0.59.0   munsell_0.5.0       
[49] zip_2.2.0            compiler_4.1.0       rlang_0.4.11.9001    grid_4.1.0           rstudioapi_0.13      igraph_1.2.6        
[55] labeling_0.4.2       rmarkdown_2.9        gtable_0.3.0         abind_1.4-5          DBI_1.1.1            rARPACK_0.11-0      
[61] curl_4.3.2           reshape2_1.4.4       R6_2.5.0             gridExtra_2.3        fastmap_1.1.0        utf8_1.2.1          
[67] stringi_1.7.3        Rcpp_1.0.7           vctrs_0.3.8          tidyselect_1.1.1     xfun_0.24
```
------------
🤔  **Expected behavior:**
If I use the following piece of code in combination with the `tune` function:
```
keepX.grid <- c(1, 3, 5, 7, ncol(exposures))
keepY.grid <- c(1, 3, 5, 7)
```
I obtain the following figure (sPLS model):
![plt](https://user-images.githubusercontent.com/22343884/126993844-fa2cf10b-7852-4cc0-b600-5f5c646e71e2.png)

with `ncol(exposures)` being indeed 20, which is shown on the Y axis in the figure. I cannot reproduce the same error using, for instance, [this tutorial](https://www.bioconductor.org/packages/release/bioc/vignettes/mixOmics/inst/doc/vignette.html#pls).

**Comment by aljabadi** [comment]
Hi @lorenzoFabbri,

Thank you for reporting this.

I tried to reproduce this using the following

```r
data(liver.toxicity)
X <- liver.toxicity$gene
Y <- liver.toxicity$clinic
set.seed(42)
tune.res = tune.spls( X, Y, ncomp = 3,
                      test.keepX = c(5, 10, 15),
                      test.keepY = c(3, 6, 8), measure = "cor",
                      folds = 5, nrepeat = 3, progressBar = TRUE)
plot(tune.res)
``` 

![image](https://user-images.githubusercontent.com/28503057/127121739-7717fa6a-8b46-4d1b-ad02-7abf928aac1e.png)


But as you can see, the order was correct. Any chance you're providing `keepX.grid` to `keepY` in the code?

**Comment by lorenzoFabbri** [comment]
Hello @aljabadi and thank you for the quick answer. Unfortunately no, I am passing the right arguments. The only way to make it work is to pass `ncol(X)` to `keepY` and `ncol(Y)` to `keepX`.

**Comment by aljabadi** [comment]
Hi @lorenzoFabbri,

My pleasure. Sorry I'm unable to reproduce this in any form using the latest github version:

``` r
data(liver.toxicity)
X <- liver.toxicity$gene
ncol(X)
#> [1] 3116
Y <- liver.toxicity$clinic
ncol(Y)
#> [1] 10
set.seed(42)
tune.res = tune.spls( X, Y, ncomp = 2,
                      test.keepX = c(5, ncol(X)),
                      test.keepY = c(3, ncol(Y)), measure = "cor",
                      folds = 5, nrepeat = 1, progressBar = TRUE)
#> 
#> tuning component: 1
#> [33m  of features tested[39m[33m[============                                      ] 25%[39m[33m  of features tested[39m[33m[=========================                         ] 50%[39m[33m  of features tested[39m[33m[=====================================             ] 75%[39m[33m  of features tested[39m[33m[==================================================] 100%[39m
#> 
#> tuning component: 2
#> [33m  of features tested[39m[33m[============                                      ] 25%[39m[33m  of features tested[39m[33m[=========================                         ] 50%[39m[33m  of features tested[39m[33m[=====================================             ] 75%[39m[33m  of features tested[39m[33m[==================================================] 100%[39m
plot(tune.res)
```

![](https://i.imgur.com/uOvPdjD.png)

<sup>Created on 2021-07-27 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

You're welcome to send us an email with the data and code so I can have a closer look.

**Comment by lorenzoFabbri** [comment]
@aljabadi I am starting to think I am not completely crazy. Here's a piece of code I retrieved from the Traceback:
```
Error in Check.entry.pls(X, Y, ncomp, keepX, keepY, mode = mode, scale = scale,  : 
  each component of 'keepX' must be no greater than 20. 
9.
stop("each component of 'keepX' must be no greater than ", P, 
    ".") 
8.
Check.entry.pls(X, Y, ncomp, keepX, keepY, mode = mode, scale = scale, 
    near.zero.var = near.zero.var, max.iter = max.iter, tol = tol, 
    logratio = logratio, DA = DA, multilevel = multilevel) 
7.
internal_wrapper.mint(X = X, Y = Y, ncomp = ncomp, scale = scale, 
    near.zero.var = near.zero.var, mode = mode, keepX = keepX, 
    keepY = keepY, max.iter = max.iter, tol = tol, logratio = logratio, 
    multilevel = multilevel, DA = FALSE, all.outputs = all.outputs) 
6.
spls(X = X, Y = Y, keepX = c(choice.keepX, test.keepX[keepX]), 
    keepY = c(choice.keepY, test.keepY[keepY]), ncomp = comp, 
    mode = mode, ...) 
5.
tune.spls(X = X, Y = Y, test.keepX = test.keepY, test.keepY = test.keepX, 
    ncomp = ncomp, validation = validation, nrepeat = nrepeat, 
    folds = folds, mode = mode, measure = measure, BPPARAM = BPPARAM, 
    progressBar = progressBar) 
4.
mixOmics::tune(method = "spls", X = exp, Y = ome, ncomp = pls.params$num.components, 
    test.keepX = eepX, test.keepY = eepY, mode = "regression", 
    nrepeat = pls.params$nrepeat, validation = "Mfold", folds = pls.params$nfolds, 
    measure = "cor", center = FALSE, scale = FALSE, max.iter = pls.params$max.iter,  ... at model.R#77
3.
run.pipeline(data = ret, pls.params = pls.params) at model.R#150
2.
pls.simple(data, pls.params, validate) at main.R#12
1.
perform.analysis.main(dat, "simple", pls.params) 
```
Precisely at point `5`. Thank you.

**Comment by aljabadi** [comment]
Thanks, @lorenzoFabbri.

It's always great to have the code which helped me realise the culprit was in fact the `tune` wrapper (and not the tune.spls function). This is now resolved in the latest GitHub version (master branch). It should be also resolved as of the release > 6.16.3 which will be built soon.
